src/persist-client/src/cli/inspect.rs

```diff
@@ -630,6 +643,204 @@ pub async fn unreferenced_blobs(args: &StateArgs) -> Result<impl serde::Serializ
+            report.num_bytes += arrow2_records
+                .updates
+                .iter()
+                .map(|r| r.goodbytes())
+                .sum::<usize>();
+
+            // Now roundtrip through arrow-rs.
+            let mut buf = Vec::new();
+            {
+                update_feature_flag("read_and_write");
+                encode_trace_parquet(&mut buf, &arrow2_records, &metrics.columnar)?;
+            }
+            let buf = SegmentedBytes::from(buf);
+
+            // Read back the arrow-rs encoded blob and make sure they also match.
+            let arrow2_records_rnd: BlobTraceBatchPart<i64> = {
+                update_feature_flag("off");
+                decode_trace_parquet(buf.clone(), &metrics.columnar)?
+            };
+            let arrow_rs_records_rnd: BlobTraceBatchPart<i64> = {
+                update_feature_flag("read_and_write");
+                decode_trace_parquet(buf.clone(), &metrics.columnar)?
+            };
+
+            if arrow2_records_rnd != arrow_rs_records_rnd {
+                anyhow::bail!("failed to roundtrip through arrow-rs, {blob_uri:?}");
+            }
+        }
+
+        report.num_shards += 1;
+        report.num_blobs += partial_keys.len();
+        tracing::info!(
+            ?blob_uri,
+            ?shard_id,
+            num_parts = partial_keys.len(),
+            "completed one check"
+        );
+    }
+
+    // Make sure our metrics indicated we did the right things.
+    let metrics = metrics_registry.gather();
+    let arrow_metrics = metrics
+        .into_iter()
+        .find(|family| family.get_name() == "mz_persist_arrow_operations")
+        .expect("failed to get metrics");
+    let get_arrow_metric = |library: &str, op: &str| {
+        arrow_metrics
+            .get_metric()
+            .iter()
+            .find(|metric| {
+                let library = metric
+                    .get_label()
+                    .iter()
+                    .find(|label| label.get_name() == "library" && label.get_value() == library)
+                    .is_some();
+                let op = metric
+                    .get_label()
+                    .iter()
+                    .find(|label| label.get_name() == "op" && label.get_value() == op)
+                    .is_some();
+
+                library && op
+            })
+            .map(|metric| {
+                let val = metric.get_counter().get_value();
+                u64::cast_lossy(val)
+            })
+            .expect("failed to find metric")
+    };
+
+    let arrow2_encode = get_arrow_metric("arrow2", "encode");
+    let arrow2_decode = get_arrow_metric("arrow2", "decode");
+    let arrow_rs_encode = get_arrow_metric("arrow_rs", "encode");
+    let arrow_rs_decode = get_arrow_metric("arrow_rs", "decode");
+
+    assert_eq!(arrow2_decode, arrow_rs_decode);
+    assert_eq!(arrow2_encode, 0);
+    assert!(arrow_rs_encode > 0);
+
+    tracing::info!(?blob_uri, "completed");
+
+    Ok(report)
```

Comment:
Let's print out the full metrics registry at the end. I think there's a helper for filtering out the uninteresting ones around here somewhere

Reply:
I couldn't find a helper (more than happy to use one though!) I added some basic filtering and logged the metrics.

I wanted to include them in the "report" but the inner `Metrics` type doesn't implement `Serialize`. I can include them in the report as a Debug formatted string if you think that's better

Reply:
Ah, sorry for the lazy comment! The one I was thinking of is https://github.com/MaterializeInc/materialize/blob/add8662444ad3af674dd9e8e24013a9a215ca368/src/persist-client/src/cli/admin.rs#L232